File: scala/anagram/src/main/scala/Anagram.scala
class Anagram(word: String) {

  lazy val lowerCaseWord = word toLowerCase
  lazy val wordChars = chars(word)

  type StringPredicate = String => Boolean

  lazy val sameChars: StringPredicate = chars(_) == wordChars
  lazy val isIdentical: StringPredicate = _.toLowerCase == lowerCaseWord
  lazy val sameCharsButNotIdentical: StringPredicate =
    candidate => sameChars(candidate) && !isIdentical(candidate)

  def matches(candidates: Seq[String]): Seq[String] =
    candidates filter sameCharsButNotIdentical

  private def chars(word: String): Seq[Char] = word.toLowerCase sorted
}

File: haskell/anagram/Anagram.hs
module Anagram (anagramsFor) where

import Data.Char (toLower)
import Data.List (sort)

anagramsFor :: String -> [String] -> [String]
anagramsFor word candidates =
  filter sameCharsButNotIdentical candidates
    where
      lowerCaseWord = toLowerCase word
      wordChars = sortedChars lowerCaseWord
      sameChars str = sortedChars str == wordChars
      isIdentical str = toLowerCase str == lowerCaseWord
      sameCharsButNotIdentical candidate = sameChars candidate && (not . isIdentical) candidate
      toLowerCase str = map toLower str
      sortedChars str = sort (toLowerCase str)